### PR TITLE
Send server logs to client using logMessage

### LINKED ISSUE #85: Send log messages to client
LSP supports sending log messages to the client, using `connection.logMessage`.
Those messages are then handled by the client in different ways, for instance VSCode displays them in the `Scalameta` output channel.

I think it would be more convenient than outputting to a file like we do now, or at least we could have both, so that one can access those logs more conveniently without leaving the editor.

### PR DESCRIPTION
**Pull request by gabro**
Closes #85

This PR implements a custom logback appender that sends server's logs down to the client.

A couple of notes on the implementation:
- every message is sent with `MessageType = Log`. The log level is included in the message itself and it's formatted consistently with the `metaserver.log` file. The `Log` message type instructs the editor (at least VSCode, @laughedelic you may want to check what Atom does) to skip the `[LEVEL - TIMESTAMP]` prefix and just print the bare message. This is way we have freedom in the formatting and we don't have to settle to a suboptimal `MessageType` conversion (for instance, there's no `MessageType.Debug`...).

- the server has to initialize the logger by setting its `connection` property. Not a brilliant design, but I think it's a bearable compromise.

- I had to tweak the VSCode client not to jump to the output automatically (otherwise it gets really annoying when using the integrated terminal. @laughedelic same as above, you may want to check the behavior on Atom)

Result (when enabling the [Output Colorizer](https://marketplace.visualstudio.com/items?itemName=IBM.output-colorizer) in VSCode):

image

### CODE REVIEW

File: metaserver/src/main/scala/scala/meta/languageserver/LSPLogger.scala [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import ch.qos.logback.classic.spi.ILoggingEvent
+import ch.qos.logback.classic.Level
+import ch.qos.logback.classic.encoder.PatternLayoutEncoder
+import ch.qos.logback.core.AppenderBase
+import langserver.core.Connection
+import langserver.types.MessageType
+import java.nio.charset.StandardCharsets.UTF_8
+import scala.beans.BeanProperty
+
+class LSPLogger(@BeanProperty var encoder: PatternLayoutEncoder)
```
Comment (gabro) [inline metaserver/src/main/scala/scala/meta/languageserver/LSPLogger.scala]: `@BeanProperty` generates `getEncoder` and `setEncoder` which are required

File: metaserver/src/main/resources/logback.groovy [hunk 2]
```diff
@@ -0,0 +1,15 @@
+def defaultPattern = "%d{HH:mm:ss.SSS} %-5level %logger{20} - %msg%n"
+
+appender("STDOUT", ch.qos.logback.core.ConsoleAppender) {
+  encoder(PatternLayoutEncoder) {
+    pattern = defaultPattern
+  }
+}
+
+appender("LSP", scala.meta.languageserver.LSPLogger) {
```
Comment (olafurpg) [inline metaserver/src/main/resources/logback.groovy]: Out of curiosity, was this not possible in logback.xml?

File: metaserver/src/main/scala/scala/meta/languageserver/LSPLogger.scala [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      if (encoder != null) new String(encoder.encode(event), UTF_8)
+      else event.getFormattedMessage
+    if (connection != null) {
+      connection.logMessage(MessageType.Log, message)
+    }
+  }
+}
+
+object LSPLogger {
+  var connection: Connection = null
```
Comment (olafurpg) [inline metaserver/src/main/scala/scala/meta/languageserver/LSPLogger.scala]: `Option[Connection]`?

File: metaserver/src/main/resources/logback.groovy (hunk 2 above)
Comment (gabro) [reply metaserver/src/main/resources/logback.groovy]: it was, but I suspect declaring re-usable variables (like `defaultPattern`) is a bit more clunky.
Overall I find the groovy configuration more readable.

File: metaserver/src/main/scala/scala/meta/languageserver/LSPLogger.scala (hunk 3 above)
Comment (gabro) [reply metaserver/src/main/scala/scala/meta/languageserver/LSPLogger.scala]: right, I went full Java 7 for a minute. Fixing this 💃

**Comment by gabro** [comment]
> We can revisit VSCode-specific parts later, when Atom supports it too.

Sure. As of now, there's nothing really VSCode specific, simply I chose to use `Log` for all the message types, which still I don't think is necessarily wrong (although I couldn't find much about it).

**Comment by gabro** [comment]
Thanks for the reviews, @laughedelic and @olafurpg. Merging! :shipit:

**Pull request merged**